hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java

```diff
@@ -80,17 +100,167 @@ public MapQueryEngineImpl(MapServiceContext mapServiceContext, QueryOptimizer op
+                // see discussion at https://github.com/hazelcast/hazelcast/pull/5049#discussion_r28773099 for details.
+                if (storedException == null) {
+                    storedException = e;
+                }
+            }
+        }
+        if (storedException != null) {
+            throw storedException;
+        }
+        return result;
+    }
+
+    private QueryResult queryParallel(String name, Predicate predicate, Collection<Integer> partitions)
+            throws InterruptedException, ExecutionException {
+        QueryResult result = newQueryResult(partitions.size());
+
+        List<Future<Collection<QueryableEntry>>> futures
+                = new ArrayList<Future<Collection<QueryableEntry>>>(partitions.size());
+
+        for (Integer partitionId : partitions) {
+            QueryPartitionCallable task = new QueryPartitionCallable(name, predicate, partitionId);
+            Future<Collection<QueryableEntry>> future = executor.submit(task);
+            futures.add(future);
+        }
+
+        Collection<Collection<QueryableEntry>> returnedResults = getResult(futures);
+        for (Collection<QueryableEntry> returnedResult : returnedResults) {
+            if (returnedResult == null) {
+                continue;
+            }
+            result.addAll(returnedResult);
+        }
+
+        return result;
+    }
+
+    private QueryResult queryParallelForPaging(String name, PagingPredicate predicate, Collection<Integer> partitions)
+            throws InterruptedException, ExecutionException {
+        QueryResult result = newQueryResult(partitions.size());
+
+        List<Future<Collection<QueryableEntry>>> futures =
+                new ArrayList<Future<Collection<QueryableEntry>>>(partitions.size());
+        for (Integer partitionId : partitions) {
+            QueryPartitionCallable task = new QueryPartitionCallable(name, predicate, partitionId);
+            Future<Collection<QueryableEntry>> future = executor.submit(task);
+            futures.add(future);
+        }
+
+        List<QueryableEntry> toMerge = new LinkedList<QueryableEntry>();
+        Collection<Collection<QueryableEntry>> returnedResults = getResult(futures);
+        for (Collection<QueryableEntry> returnedResult : returnedResults) {
+            toMerge.addAll(returnedResult);
+        }
+
+        Map.Entry<Integer, Map.Entry> nearestAnchorEntry = getNearestAnchorEntry(predicate);
+        List<QueryableEntry> sortedSubList = getSortedSubList(toMerge, predicate, nearestAnchorEntry);
+        result.addAll(sortedSubList);
+        return result;
+    }
+
+    private static Collection<Collection<QueryableEntry>> getResult(List<Future<Collection<QueryableEntry>>> lsFutures) {
+        return returnWithDeadline(lsFutures, QUERY_EXECUTION_TIMEOUT_MINUTES, MINUTES, RETHROW_EVERYTHING);
+    }
+
+    private boolean checkPartitionState(int expectedPartitionStateVersion, Predicate predicate) {
```

Comment:
I always struggled with the meaning of this method (due to my missing knowledge what a partition state is and the not very helpful method name). What about `isPartitionStateUnchanged` or `hasPartitionStateChanged` as method name? That would at least make clear what `true` and `false` means.